receiver/splunkhecreceiver/receiver.go

```diff
@@ -0,0 +1,305 @@
+		return nil, errEmptyEndpoint
+	}
+
+	r := &splunkReceiver{
+		logger:       logger,
+		config:       &config,
+		nextConsumer: nextConsumer,
+		server: &http.Server{
+			Addr: config.Endpoint,
+			// TODO: Evaluate what properties should be configurable, for now
+			//		set some hard-coded values.
+			ReadHeaderTimeout: defaultServerTimeout,
+			WriteTimeout:      defaultServerTimeout,
+		},
+	}
+
+	mx := mux.NewRouter()
+	mx.HandleFunc(hecPath, r.handleReq)
+	r.server.Handler = mx
+
+	return r, nil
+}
+
+// StartMetricsReception tells the receiver to start its processing.
+// By convention the consumer of the received data is set when the receiver
+// instance is created.
+func (r *splunkReceiver) Start(_ context.Context, host component.Host) error {
+	r.Lock()
+	defer r.Unlock()
+
+	err := componenterror.ErrAlreadyStarted
+	r.startOnce.Do(func() {
+		err = nil
+
+		go func() {
+			if r.config.TLSCredentials != nil {
+				host.ReportFatalError(r.server.ListenAndServeTLS(r.config.TLSCredentials.CertFile, r.config.TLSCredentials.KeyFile))
+			} else {
+				if err = r.server.ListenAndServe(); err != nil && err != http.ErrServerClosed {
+					host.ReportFatalError(err)
+				}
+			}
+		}()
+	})
+
+	return err
+}
+
+// StopMetricsReception tells the receiver that should stop reception,
+// giving it a chance to perform any necessary clean-up.
+func (r *splunkReceiver) Shutdown(context.Context) error {
+	r.Lock()
+	defer r.Unlock()
+
+	err := componenterror.ErrAlreadyStopped
+	r.stopOnce.Do(func() {
+		err = r.server.Close()
+	})
+	return err
+}
+
+func (r *splunkReceiver) handleReq(resp http.ResponseWriter, req *http.Request) {
+	transport := "http"
+	if r.config.TLSCredentials != nil {
+		transport = "https"
+	}
+	ctx := obsreport.ReceiverContext(req.Context(), r.config.Name(), transport, r.config.Name())
+	ctx = obsreport.StartMetricsReceiveOp(ctx, r.config.Name(), transport)
+
+	if req.Method != http.MethodPost {
+		r.failRequest(ctx, resp, http.StatusBadRequest, invalidMethodRespBody, nil)
+		return
+	}
+
+	if req.Header.Get(httpContentTypeHeader) != jsonContentType {
+		r.failRequest(ctx, resp, http.StatusUnsupportedMediaType, invalidContentRespBody, nil)
+		return
+	}
+
+	encoding := req.Header.Get(httpContentEncodingHeader)
+	if encoding != "" && encoding != gzipEncoding {
+		r.failRequest(ctx, resp, http.StatusUnsupportedMediaType, invalidEncodingRespBody, nil)
+		return
+	}
+
+	bodyReader := req.Body
+	if encoding == gzipEncoding {
+		var err error
+		bodyReader, err = gzip.NewReader(bodyReader)
+		if err != nil {
+			r.failRequest(ctx, resp, http.StatusBadRequest, errGzipReaderRespBody, err)
+			return
+		}
+	}
+
+	body, err := ioutil.ReadAll(bodyReader)
+	if err != nil {
+		r.failRequest(ctx, resp, http.StatusBadRequest, errReadBodyRespBody, err)
+		return
+	}
+	if len(body) == 0 {
+		resp.Write(okRespBody)
+		return
+	}
+
+	messagesReceived := 0
+	for _, payload := range bytes.Split(body, []byte("\r\n\r\n")) {
```

Comment:
Perhaps you can use json decoder instead of splitting? https://golang.org/pkg/encoding/json/#Decoder Didn't test with \r but at least with \n separator it seems to work:

```go
func main() {
	const jsonStream = `{"Name": "Ed", "Text": "Knock knock."}


		{"Name": "Sam", "Text": "Who's there?"}
		{"Name": "Ed", "Text": "Go fmt."}


		{"Name": "Sam", "Text": "Go fmt who?"}
		{"Name": "Ed", "Text": "Go fmt yourself!"}`
	type Message struct {
		Name, Text string
	}
	dec := json.NewDecoder(strings.NewReader(jsonStream))

	// while the array contains values
	for dec.More() {
		var m Message
		// decode an array value (Message)
		err := dec.Decode(&m)
		if err != nil {
			log.Fatal(err)
		}

		fmt.Printf("%v: %v\n", m.Name, m.Text)
	}
}
```